### PR TITLE
ARROW-1533: [JAVA] realloc should consider the existing buffer capacity for computing target memory requirement

### PR DESCRIPTION
**Pull request by siddharthteotia**
cc @jacques-n , @StevenMPhillips 

Problem description:

We recently encountered a problem when we were trying to add JSON files with complex schema as datasets.

Initially we started with a Float8Vector with default memory allocation of (4096 * 8) 32KB.
Went through several iterations of setSafe() to trigger a realloc() from 32KB to 64KB.

Another round of setSafe() calls to trigger a realloc() from 64KB to 128KB

After that we encountered a BigInt and promoted our vector to UnionVector. This required us to create a UnionVector with BigIntVector and Float8Vector. The latter required us to transfer the Float8Vector we were earlier working with to the Float8Vector inside the Union.

As part of transferTo(), the target Float8Vector got all the ArrowBuf state (capacity, buffer contents) etc transferred from the source vector.

Later, a realloc was triggered on the Float8Vector inside the UnionVector.

The computation inside realloc() to determine the amount of memory to be reallocated goes wrong since it makes the decision based on allocateSizeInBytes – although this vector was created as part of transfer() from 128KB source vector, allocateSizeInBytes is still at the initial/default value of 32KB

We end up allocating a 64KB buffer and attempt to copy 128KB over 64KB and seg fault when invoking setBytes() whereas our assumption was we will be reallocating to 256KB.

There is a wrong assumption in realloc() that allocateSizeInBytes is always equal to data.capacity(). The particular scenario described above exposes where this assumption could go wrong.

### CODE REVIEW

**Comment by wesm** [comment]
Since we have a 0.7.0 incoming shortly, I want to make sure that this is not a blocker. We can always release 0.7.1 with such bug fixes shortly thereafter.

File: java/vector/src/main/codegen/templates/FixedValueVectors.java [hunk 1]
```diff
@@ -208,14 +208,19 @@ private void allocateBytes(final long size) {
...
    */
   public void reAlloc() {
-    final long newAllocationSize = allocationSizeInBytes * 2L;
-    if (newAllocationSize > MAX_ALLOCATION_SIZE)  {
+    long baseSize  = allocationSizeInBytes;
+    final int currentBufferCapacity = data.capacity();
+    if (baseSize < (long)currentBufferCapacity) {
+        baseSize = (long)currentBufferCapacity;
+    }
+    final long newAllocationSize = baseSize * 2L;
```
Comment (jacques-n) [inline java/vector/src/main/codegen/templates/FixedValueVectors.java]: we should make sure this is a power of 2.

**Review by jacques-n — COMMENTED** [review]
Change makes sense. We should make sure the allocation is a power of two.

**Comment by siddharthteotia** [comment]
addressed review comments

**Comment by siddharthteotia** [comment]
Added unit tests.

**Comment by siddharthteotia** [comment]
Without the code-changes, the unit test fails with segmentation fault (as described in the problem above)

Tests run: 21, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.209 s <<< FAILURE! - in org.apache.arrow.vector.TestValueVector
testReallocAfterVectorTransfer(org.apache.arrow.vector.TestValueVector)  Time elapsed: 0.045 s  ERROR!
java.lang.IndexOutOfBoundsException: index: 0, length: 131072 (expected: range(0, 65536))
org.apache.arrow.vector.TestValueVector.testReallocAfterVectorTransfer(TestValueVector.java:962)

File: java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java [hunk 2]
```diff
@@ -889,10 +890,94 @@ public void testNullableVarType2() {
    *  TODO:
    *
    *  The realloc() related tests below should be moved up and we need to
-   *  realloc related tests (edge cases) for more vector types.
+   *  add realloc related tests (edge cases) for more vector types.
    */
 
   @Test
+  public void testReallocAfterVectorTransfer() {
```
Comment (jacques-n) [inline java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java]: Can you add tests for a NullableBitVector and NullableVarCharVector? Does MapVector or ListVector also need a test?

File: java/vector/src/main/java/org/apache/arrow/vector/BitVector.java [hunk 3]
```diff
@@ -208,15 +208,22 @@ private void allocateBytes(final long size) {
...
    */
   public void reAlloc() {
-    final long newAllocationSize = allocationSizeInBytes * 2L;
+    long baseSize  = allocationSizeInBytes;
+    final int currentBufferCapacity = data.capacity();
+    if (baseSize < (long)currentBufferCapacity) {
+      baseSize = (long)currentBufferCapacity;
+    }
+    long newAllocationSize = baseSize * 2L;
+    newAllocationSize = Long.highestOneBit(newAllocationSize - 1) << 1;
```
Comment (jacques-n) [inline java/vector/src/main/java/org/apache/arrow/vector/BitVector.java]: Maybe use BaseAllocator.nextPowerOfTwo()

**Comment by siddharthteotia** [comment]
@jacques-n , added tests for fixed-width vector, variable width vector and bit vector. To further test correctness, I verified that without the code changes, unit tests fail with expected segmentation fault in reAlloc().

At this point, I don't think we need tests for other vector types for the problem this patch is fixing. 

However, I have two test related patches in pipeline 

(1) tests for mapvector -- we don't have any unit tests for MapVector
(2) splitAndTransfer for all vector types -- we recently added for bit vector, union vector, list vector. Need to cover other types too.

I will ensure that those patches cover tests for this scenario as well.

**Comment by siddharthteotia** [comment]
The unit tests added here cover both nullable and non-nullable vectors.

**Pull request closed without merging**